Fix `Nullable<bool>.ToString()` conversion

ranma42 opened the pull request:
This PR updates translation of `bool?` expressions in 2 ways:
 - `null` is now converted into the empty string, matching the usual C# behavior
 - expressions are assumed to be nullable (they were previously assumed to be non-nullable)

Fixes #33941 for SQLite.
Contributes to #33942.

maumar commented:
There are some product changes here. Consider splitting them out to make this pr test only, alternatively pull in the product changes for https://github.com/dotnet/efcore/pull/33938 into one pr and update the PR description. Changes look good though.

ranma42 commented:
@maumar I moved the test updates and dataset changes in
 - https://github.com/dotnet/efcore/pull/34019

I am not 100% sure if this is the "product changes"/all of them; I did not include the tests that were being added as I thought that they might belong to the specific fix/changeset.
I rebased this PR on top of that and marked it as draft as it is basically dependent on it.

Note that in order to ensure that the test suite passes, I disabled `ToString_boolean_property_nullable` in the PR the base PR and re-enabled it here.

maumar commented:
@ranma42 huge apologies! I wanted to resolve conflicts in https://github.com/dotnet/efcore/pull/34014 before approval and merge, and I accidentally force pushed into this branch instead (auto-complete >.<) - please tell me you still have these changes locally on your machine...

ranma42 commented:
> @ranma42 huge apologies! I wanted to resolve conflicts in #34014 before approval and merge, and I accidentally force pushed into this branch instead (auto-complete >.<) - please tell me you still have these changes locally on your machine...

yup :) rebased and pushed this one 👍 

also, in case it happens with your own commits (and/or with branches you have not replicated elsewhere, remember that `git` defaults to keeping commits around for 90 days; `git reflog` is very effective in saving you from this kind of mistakes

maumar commented:
phew! Thanks making the PR (twice! ;) )
